i2c: stm32: RTIO: add target mode and transfers cases without START/STOP in between

etienne-lms opened the pull request:
Implement target mode in STM32 I2C v1 and v2 drivers for RTIO. The target mode implementation is dumped from the non-RTIO related drivers (*i2c_ll_stm32_v1.c* and *i2c_ll_stm32_v2.c*).

Fix write request issues found testing against *i2c/rtio_loopback* sample.

This series also adds 2 STM32 boards to the *i2c/rtio_loopback* sample: one based on STM32 I2C v1 and one based on STM32 I2C v2.

etienne-lms committed:
drivers: i2c: stm32: I2C transaction first transfer must set RESTART flag

The first transfer in an I2C transaction series must always set
the RESTART flag. Since this is not always done by callers,
enforce the flag is set, both in the I2C API and RTIO API entry
points of STM32 I2C RTIO drivers.

Add a sanity test in STM32 I2C v2 RTIO driver on transaction
chunks regarding RESTART and STOP flags, as done on the non-RTIO
driver.


Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: i2c: stm32: add target mode in RTIO drivers

Implement I2C target mode in STM32 I2C RTIO drivers. The implementation
in respectively i2c_ll_stm32_v1_rtio.c and i2c_ll_stm32_v2_rtio.c is
based on the implementation of the non-RTIO drivers, respectively
i2c_ll_stm32_v1.c and i2c_ll_stm32_v2_rtio.c.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
samples: drivers: i2c: rtio_loopback: add nucleo_f401re/nucleo_h503rb

Add nucleo_f401re and nucleo_h503rb boards to rtio_loopback tests
on the I2C driver.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: i2c: stm32: report error case in v1 RTIO driver

Report error event to the RTIO framework in STM32 I2C v1 driver.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

teburd commented:
I'll take a look friday, this is something I've been meaning to try myself and maybe work out some more details on how the target mode should work (as it seems like the target mode callbacks vary slightly depending on the vendor...)

etienne-lms committed:
drivers: i2c: stm32: fix write burst in RTIO drivers

Correct management of I2C write transfers that are driven from 2
I2C messages, one for the transfer of the address to write
to and a following message for the transfer of the data to be written.
In this case, no START and STOP bit should be triggered between the 2
I2C transfers.

In STM32 I2C v1 driver, this simply means not emitting a START
condition when the input sequence does not set the RESTART flag.

In STM32 I2C v2 driver, this can only be achieved using the SoC I2C
controller Reload Mode feature, hence update Reload Mode implementation
to use reload mode for both transfers over 255 bytes (splitted in
chunks) and for cases where no STOP and START bits are emitted between
2 transfer in the same transaction.

Regarding STM32 I2C v2 driver, the reload mode must be enable before the
first transaction, known that it's needed due to the following
transaction. Therefore we need to parse the messages grouped in the same
transaction to detect such sequences (a message without STOP flag
followed by a message without RESTART flag) and when so, pass the
information to the I2C driver through RTIO framework. To achieve that,
let's use a free bit from I2C message flags (bit 7) in the message
for which we need to enable SoC ReloadMode from the first transfer.
We check that this bit flag is not used by the I2C framework prior
using it.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

teburd commented:
Looks good to me, thank you!
